lib/media/time_ranges_utils.js

```diff
@@ -155,4 +155,46 @@ shaka.media.TimeRangesUtils = class {
     }
     return ret;
   }
+
+  /**
+   * This operation can be potentially EXPENSIVE and should only be done in
+   * debug builds for debugging purposes.
+   *
+   * @param {TimeRanges} oldRanges
+   * @param {TimeRanges} newRanges
+   * @return {?shaka.extern.BufferedRange}
+   */
+  static computeAddedRanges(oldRanges, newRanges) {
+    const TimeRangesUtils = shaka.media.TimeRangesUtils;
+
+    if (!oldRanges || !oldRanges.length) {
+      return null;
+    }
+    if (!newRanges || !newRanges.length) {
+      return TimeRangesUtils.getBufferedInfo(newRanges).pop();
+    }
+
+    const added = [];
+    for (const newRange of TimeRangesUtils.getBufferedInfo(newRanges)) {
+      let foundOverlap = false;
```

Comment:
Perhaps I'm not thinking of some possible scenario here, but couldn't this algorithm potentially be a little faster for super-messy inputs if you read `newRanges` and `oldRanges` in reverse order, and returned if you saw a non-overlap or a newly-added difference? That would return early in some circumstances, and would remove the need to keep an array of added ranges. Something like...

```js
    const reversedNewRanges = TimeRangesUtils.getBufferedInfo(newRanges).reverse();
    const reversedOldRanges = TimeRangesUtils.getBufferedInfo(oldRanges).reverse();
    for (const newRange of reversedNewRanges) {
      let foundOverlap = false;
      for (const oldRange of reversedOldRanges) {
        if (oldRange.end >= newRange.start && oldRange.end <= newRange.end) {
          foundOverlap = true;

          // If the new range goes beyond the corresponding old one, the
          // difference is newly-added.
          if (newRange.end > oldRange.end) {
            return {start: oldRange.end, end: newRange.end};
          }
        }
      }

      if (!foundOverlap) {
        return newRange;
      }
    }

    return null;
```

Reply:
You're right.  I started off trying to build something very general, then gave up and hacked it to work better for our purposes.  Your way is better.

Reply:
Done